library/std/src/sys/sync/rwlock/queue.rs

```diff
@@ -456,25 +484,83 @@ impl RwLock {
     /// * There must be threads queued on the lock.
     #[cold]
     unsafe fn unlock_contended(&self, mut state: State) {
+        debug_assert!(state.addr() & STATE == (QUEUED | LOCKED));
         loop {
             // Atomically release the lock and try to acquire the queue lock.
-            let next = state.map_addr(|a| (a & !LOCKED) | QUEUE_LOCKED);
-            match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
-                // The queue lock was acquired. Release it, waking up the next
-                // waiter in the process.
-                Ok(_) if state.addr() & QUEUE_LOCKED == 0 => unsafe {
-                    return self.unlock_queue(next);
-                },
+            let next = state.wrapping_byte_sub(LOCKED);
+            if state.addr() & QUEUE_LOCKED != 0 {
                 // Another thread already holds the queue lock, leave waking up
                 // waiters to it.
-                Ok(_) => return,
-                Err(new) => state = new,
+                match self.state.compare_exchange_weak(state, next, Release, Relaxed) {
+                    Ok(_) => return,
+                    Err(new) => state = new,
+                }
+            } else {
+                // Acquire the queue lock and wake up the next waiter.
+                let next = next.wrapping_byte_add(QUEUE_LOCKED);
+                match self.state.compare_exchange_weak(state, next, AcqRel, Relaxed) {
+                    Ok(_) => unsafe {
+                        return self.unlock_queue(next);
+                    },
+                    Err(new) => state = new,
+                }
             }
         }
     }
 
-    /// Unlocks the queue. If the lock is unlocked, wakes up the next eligible
-    /// thread(s).
+    #[inline]
+    pub unsafe fn downgrade(&self) {
+        // Atomically set to read-locked with a single reader, without any
+        // waiting threads.
+        if let Err(state) = self.state.compare_exchange(
+            without_provenance_mut(LOCKED),
+            without_provenance_mut(SINGLE | LOCKED),
+            Release,
+            Relaxed,
+        ) {
+            // The only way the state can have changed is if there are threads
+            // queued. Wake all of them up.
+            unsafe { self.downgrade_slow(state) }
+        }
+    }
+
+    #[cold]
+    unsafe fn downgrade_slow(&self, mut state: State) {
```

Comment:
This function does a bit, can it get a doc comment?

Reply:
I added:
```rust
    /// Downgrades the lock from write-locked to read-locked in the case that there are threads
    /// waiting on the wait queue.
    ///
    /// This function will either wake up all of the waiters on the wait queue or designate the
    /// current holder of the queue lock to wake up all of the waiters instead. Once the waiters
    /// wake up, they will continue in the execution loop of `lock_contended`.
```
I think this makes sense, but I don't know if it's specific enough.